src/code/z_vimode.c

```diff
@@ -1,23 +1,346 @@
+    hiResInterlaced = hiRes && modeN;
+    loResDeinterlaced = loRes && modeN;
+    loResInterlaced = loRes && modeF;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_vimode/func_80140900.s")
+    modeLAN1 = loRes && antialiasOn && modeN && fb16Bit;
+    modeLPN2 = loRes && antialiasOff && modeN && fb32Bit;
+    modeHPN2 = hiRes && antialiasOff && modeN && fb32Bit;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_vimode/func_8014090C.s")
+    upperAdjust &= ~1;
+    lowerAdjust &= ~1;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_vimode/func_80140970.s")
+    yScaleLo = (hiResDeflicker ? 2 : 1) * (((SCREEN_HEIGHT << 11) / SCREEN_HEIGHT) / (loRes ? 1 : 2));
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_vimode/func_80140CE0.s")
+    yScaleHiEvenField = modeF ? (loResInterlaced ? 0x1000000 : 0x2000000) : 0;
+    yScaleHiOddField = modeF ? (loResInterlaced ? 0x3000000 : 0x2000000) : 0;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_vimode/func_80140D04.s")
+    viMode->type = type;
+    viMode->comRegs.ctrl = OS_VI_UNK2000 | OS_VI_UNK1000 | OS_VI_GAMMA | OS_VI_GAMMA_DITHER |
+                           (!loResDeinterlaced ? OS_VI_UNK40 : 0) | (antialiasOn ? OS_VI_DIVOT : 0) |
+                           (fb32Bit ? OS_VI_UNK2 | OS_VI_UNK1 : OS_VI_UNK2);
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_vimode/func_80140D10.s")
+    if (modeLAN1) {
+        // Anti-aliased, fetch extra lines as-needed
+        viMode->comRegs.ctrl |= 0x100;
+    } else if (modeLPN2 | modeHPN2) {
+        // Point-sampled, resampling disabled
+        viMode->comRegs.ctrl |= 0x300;
+    } else {
+        if (antialiasOff) {
+            // Point-sampled, resampling enabled
+            viMode->comRegs.ctrl |= 0x200;
+        } else {
+            // Anti-aliased, always fetch extra lines
+            viMode->comRegs.ctrl |= 0;
+        }
+    }
+
+    viMode->comRegs.width = width * (hiResInterlaced ? 2 : 1);
+
+    if (tvType < 3) {
+        static const ViModeStruct D_801DF0C0[] = {
+            {
+                // OS_TV_PAL
+                BURST(58, 35, 4, 64),
+                VSYNC(624),
+                HSYNC(3177, 21),
+                LEAP(3183, 3182),
+                HSTART(128, 768),
+                START(95, 569),
+                BURST(107, 2, 9, 0),
+            },
+            {
+                // OS_TV_NTSC
+                BURST(57, 34, 5, 62),
+                VSYNC(524),
+                HSYNC(3093, 0),
+                LEAP(3093, 3093),
+                HSTART(108, 748),
+                START(37, 511),
+                BURST(4, 2, 14, 0),
+            },
+            {
+                // OS_TV_MPAL
+                BURST(57, 30, 5, 70),
+                VSYNC(524),
+                HSYNC(3088, 0),
+                LEAP(3100, 3100),
+                HSTART(108, 748),
+                START(37, 511),
+                BURST(4, 2, 14, 0),
+            },
+        };
+        const ViModeStruct* ptr = &D_801DF0C0[tvType];
```

Comment:
Does this data need to be in-function to match? It breaks of the flow of the function

Reply:
Yes. It's a refactoring from OoT that eliminates an if statement